### PR TITLE
[MTDSA-26988] submit final declatration CalculationType path parameter

### PR DESCRIPTION
**Pull request by juamalBlackman**

### CODE REVIEW

**Review by GazalNebHMRC — CHANGES_REQUESTED** [review]

File: app/api/errors/IndividualCalculationsErrors.scala [hunk 1]
```diff
@@ -33,6 +33,20 @@ object RuleIncomeSourcesChangedError
       BAD_REQUEST
     )
 
+object RuleOutsideAmendmentWindowError
+    extends MtdError(
+      code = "RULE_OUTSIDE_AMENDMENT_WINDOW",
+      message = "The remote endpoint has indicated that the request cannot be completed as you are outside the amendment window.",
```
[the remark this answers is not in the document]
Comment (Tateni93) [reply app/api/errors/IndividualCalculationsErrors.scala:39]: https://jira.tools.tax.service.gov.uk/browse/MTDSA-26988 - This is straight out of the ticket. I'll check anyway

[the remark this answers is not in the document]
Comment (Tateni93) [reply app/api/errors/IndividualCalculationsErrors.scala:39]: Asked Sion to take a look. Will update on his response.

File: test/v7/submitFinalDeclaration/def1/Def1_SubmitFinalDeclarationRulesValidatorSpec.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import api.errors.RuleSubmissionFailedError
+import cats.data.Validated
+import cats.data.Validated.{Invalid, Valid}
+import shared.models.domain.{CalculationId, Nino, TaxYear}
+import shared.models.errors.MtdError
+import shared.utils.UnitSpec
+import v7.submitFinalDeclaration.model.request.Def1_SubmitFinalDeclarationRequestData
+import v7.submitFinalDeclaration.model.request.domain.{CalculationType, `confirm-amendment`}
+
+class Def1_SubmitFinalDeclarationRulesValidatorSpec extends UnitSpec {
```
[the remark this answers is not in the document]
Comment (Tateni93) [reply test/v7/submitFinalDeclaration/def1/Def1_SubmitFinalDeclarationRulesValidatorSpec.scala:28]: I disagree here, the Rules Validator is specific to determining whether a valid request satisfies business logic making it different from validating the format of path parameters and is its own object therefore a separate spec is required.

File: app/api/errors/IndividualCalculationsErrors.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (sion-docs) [reply app/api/errors/IndividualCalculationsErrors.scala:39]: See the first point here - https://confluence.tools.tax.service.gov.uk/display/MTE/How+to+write+error+messages. I suggest changing to "The request cannot be completed as you are outside the amendment window"

**Comment by hmrc-web-operations** [comment]
Thank you for submitting a pull request on this HMRC repository. A member of the HMRC team will review your changes and action accordingly.

HMRC reviewer: please reference [this runbook](https://github.com/hmrc/build-and-deploy/blob/main/runbooks/reviewing_external_pull_requests.md) for guidance on how to proceed.

File: app/v7/submitFinalDeclaration/SubmitFinalDeclarationValidatorFactory.scala [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+@Singleton
+class SubmitFinalDeclarationValidatorFactory {
+
+  def validator(nino: String,
+                taxYear: String,
+                calculationId: String,
+                calculationType: String,
+                schema: SubmitFinalDeclarationSchema): Validator[SubmitFinalDeclarationRequestData] =
+    schema match {
+      case Def1 => new Def1_SubmitFinalDeclarationValidator(nino, taxYear, calculationId, calculationType)
```
Comment (EdCope) [inline app/v7/submitFinalDeclaration/SubmitFinalDeclarationValidatorFactory.scala:35]: The import for `Def1` is missing here

File: resources/public/api/conf/7.0/common/errors.yaml [hunk 4]
```diff
@@ -25,6 +25,20 @@ components:
...
         message: The provided tax year is invalid
 
+    formatCalculationType:
+      description: |
+        The format of the supplied calculationType value is not valid.
+      value:
+        code: FORMAT_CALCULATION_TYPE
+        message: The provided calculationType is not invalid
+
+    ruleOutisdeAmendmentWindow:
```
Comment (EdCope) [inline resources/public/api/conf/7.0/common/errors.yaml:35]: There is a typo here `...OutisdeAme...`

**Comment by hmrc-web-operations** [comment]
Thank you for submitting a pull request on this HMRC repository. A member of the HMRC team will review your changes and action accordingly.

HMRC reviewer: please reference [this runbook](https://github.com/hmrc/build-and-deploy/blob/main/runbooks/reviewing_external_pull_requests.md) for guidance on how to proceed.

File: resources/public/api/conf/7.0/submit_final_declaration.yaml [hunk 5]
```diff
@@ -60,6 +62,10 @@ post:
           examples:
             FORMAT_NINO:
               $ref: ./common/errors.yaml#/components/examples/formatNino
+            FORMAT_CALCULATION_TYPE:
+              $ref: ./common/errors.yaml#/components/examples/formatCalculationType
+            RULE_OUTSIDE_AMENDMENT_WINDOW:
+              $ref: ./common/errors.yaml#/components/examples/ruleOutisdeAmendmentWindow
```
Comment (EdCope) [inline resources/public/api/conf/7.0/submit_final_declaration.yaml:68]: Use the same typo correction here

**Comment by hmrc-web-operations** [comment]
Thank you for submitting a pull request on this HMRC repository. A member of the HMRC team will review your changes and action accordingly.

HMRC reviewer: please reference [this runbook](https://github.com/hmrc/build-and-deploy/blob/main/runbooks/reviewing_external_pull_requests.md) for guidance on how to proceed.

File: app/v7/submitFinalDeclaration/def1/Def1_SubmitFinalDeclarationRulesValidator.scala [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+import v7.submitFinalDeclaration.model.request.Def1_SubmitFinalDeclarationRequestData
+import v7.submitFinalDeclaration.model.request.domain.`confirm-amendment`
+
+object Def1_SubmitFinalDeclarationRulesValidator extends RulesValidator[Def1_SubmitFinalDeclarationRequestData] {
+
+  def validateBusinessRules(parsed: Def1_SubmitFinalDeclarationRequestData): Validated[Seq[MtdError], Def1_SubmitFinalDeclarationRequestData] = {
+
+    def validatedCalculationType(request: Def1_SubmitFinalDeclarationRequestData): Validated[Seq[MtdError], Unit] ={
+      (request.calculationType) match {
+        case `confirm-amendment` if request.taxYear.year < 2026 => Invalid(Seq(RuleSubmissionFailedError))
```
[the remark this answers is not in the document]
Comment (Tateni93) [reply app/v7/submitFinalDeclaration/def1/Def1_SubmitFinalDeclarationRulesValidator.scala:33]: I'm not sure what is meant here? Are we saying the current message attached to this error needs to be overridden for this case?

**Comment by hmrc-web-operations** [comment]
Thank you for submitting a pull request on this HMRC repository. A member of the HMRC team will review your changes and action accordingly.

HMRC reviewer: please reference [this runbook](https://github.com/hmrc/build-and-deploy/blob/main/runbooks/reviewing_external_pull_requests.md) for guidance on how to proceed.

File: app/api/errors/IndividualCalculationsErrors.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (GazalNebHMRC) [reply app/api/errors/IndividualCalculationsErrors.scala]: Can you please update the error message to ""The request cannot be completed as you are outside the amendment window" as indicated above? Also, @sion-docs Can you please update the business spec [API#2082(1780) - Submit a Self Assessment Final Declaration and Confirm Amendment (TYS)](https://confluence.tools.tax.service.gov.uk/pages/viewpage.action?pageId=955810017) with the new error message for RULE_OUTSIDE_AMENDMENT_WINDOW as well?

File: app/v7/submitFinalDeclaration/def1/Def1_SubmitFinalDeclarationRulesValidator.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (GazalNebHMRC) [reply app/v7/submitFinalDeclaration/def1/Def1_SubmitFinalDeclarationRulesValidator.scala:33]: After confirmation with our software architect Fatih, it seems no further change is required for this, thanks

File: test/v7/submitFinalDeclaration/SubmitFinalDeclarationServiceSpec.scala [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            .submitFinalDeclaration(request)
+            .returns(Future.successful(Left(ResponseWrapper(correlationId, DownstreamErrors.single(DownstreamErrorCode(downstreamErrorCode))))))
+
+          val result: Either[ErrorWrapper, ResponseWrapper[Unit]] = await(service.submitFinalDeclaration(request))
+          result shouldBe Left(ErrorWrapper(correlationId, error))
+        }
+
+      val errors = List(
+        ("INVALID_TAXABLE_ENTITY_ID", NinoFormatError),
+        ("INVALID_TAX_YEAR", TaxYearFormatError),
```
[the remark this answers is not in the document]
Comment (Tateni93) [reply test/v7/submitFinalDeclaration/SubmitFinalDeclarationServiceSpec.scala:79]: INVALID_CALCULATION_ID  is on Line 80 but I have added Calculation Type and Outside Amendment Window tests

**Comment by hmrc-web-operations** [comment]
Thank you for submitting a pull request on this HMRC repository. A member of the HMRC team will review your changes and action accordingly.

HMRC reviewer: please reference [this runbook](https://github.com/hmrc/build-and-deploy/blob/main/runbooks/reviewing_external_pull_requests.md) for guidance on how to proceed.

**Comment by hmrc-web-operations** [comment]
Thank you for submitting a pull request on this HMRC repository. A member of the HMRC team will review your changes and action accordingly.

HMRC reviewer: please reference [this runbook](https://github.com/hmrc/build-and-deploy/blob/main/runbooks/reviewing_external_pull_requests.md) for guidance on how to proceed.

**Comment by hmrc-web-operations** [comment]
Thank you for submitting a pull request on this HMRC repository. A member of the HMRC team will review your changes and action accordingly.

HMRC reviewer: please reference [this runbook](https://github.com/hmrc/build-and-deploy/blob/main/runbooks/reviewing_external_pull_requests.md) for guidance on how to proceed.

File: app/api/errors/IndividualCalculationsErrors.scala [hunk 8]
```diff
@@ -33,6 +33,20 @@ object RuleIncomeSourcesChangedError
       BAD_REQUEST
     )
 
+object RuleOutsideAmendmentWindowError
+    extends MtdError(
+      code = "RULE_OUTSIDE_AMENDMENT_WINDOW",
+      message = "You are outside the amendment window",
```
Comment (EdCope) [inline app/api/errors/IndividualCalculationsErrors.scala:39]: This comment was on before you rebased, `The request cannot be completed as you are outside the amendment window`

Comment (Tateni93) [reply app/api/errors/IndividualCalculationsErrors.scala:39]: Sion messaged me again this morning to say that the message should be You are Outside the amendment window. @sion-docs is this still the correct message?

File: app/v7/submitFinalDeclaration/model/request/domain/CalculationType.scala
[the remark this answers is not in the document]
Comment (Tateni93) [reply app/v7/submitFinalDeclaration/model/request/domain/CalculationType.scala:1]: I can't see that path under common folder there is a shared.models.domain. Do you mean to move it there?

File: app/api/errors/IndividualCalculationsErrors.scala [hunk 9]
```diff
@@ -33,6 +33,20 @@ object RuleIncomeSourcesChangedError
...
+      BAD_REQUEST
+    )
+
+object FormatCalculationTypeError
+    extends MtdError(
+      code = "FORMAT_CALCULATION_TYPE",
+      message = "The provided calculation type is not valid.",
```
Comment (EdCope) [inline app/api/errors/IndividualCalculationsErrors.scala:46]: Please remove the full stop for the message

File: resources/public/api/conf/7.0/common/errors.yaml [hunk 10]
```diff
@@ -25,6 +25,20 @@ components:
...
+      value:
+        code: FORMAT_CALCULATION_TYPE
+        message: The provided calculationType is not invalid
+
+    ruleOutsideAmendmentWindow:
+      description: |
+        The client and/or agent is outside the amendment window.
+      value:
+        code: RULE_OUTSIDE_AMENDMENT_WINDOW
+        message: You are outside the amendment window
```
Comment (EdCope) [inline resources/public/api/conf/7.0/common/errors.yaml:40]: Match to above when we have received confirmation for your question:

https://github.com/hmrc/individual-calculations-api/pull/545/files#r1821198876

**Comment by hmrc-web-operations** [comment]
Thank you for submitting a pull request on this HMRC repository. A member of the HMRC team will review your changes and action accordingly.

HMRC reviewer: please reference [this runbook](https://github.com/hmrc/build-and-deploy/blob/main/runbooks/reviewing_external_pull_requests.md) for guidance on how to proceed.

**Comment by hmrc-web-operations** [comment]
Thank you for submitting a pull request on this HMRC repository. A member of the HMRC team will review your changes and action accordingly.

HMRC reviewer: please reference [this runbook](https://github.com/hmrc/build-and-deploy/blob/main/runbooks/reviewing_external_pull_requests.md) for guidance on how to proceed.

**Comment by hmrc-web-operations** [comment]
Thank you for submitting a pull request on this HMRC repository. A member of the HMRC team will review your changes and action accordingly.

HMRC reviewer: please reference [this runbook](https://github.com/hmrc/build-and-deploy/blob/main/runbooks/reviewing_external_pull_requests.md) for guidance on how to proceed.

File: app/api/errors/IndividualCalculationsErrors.scala (hunk 8 above)
Comment (sion-docs) [reply app/api/errors/IndividualCalculationsErrors.scala:39]: Correct, it should be simply "You are outside the amendment window". This is for consistency with other error messages, and 'The request cannot be completed' is redundant as they are receiving an error message.

**Review by sion-docs — DISMISSED** [review]

File: app/api/errors/IndividualCalculationsErrors.scala (hunk 1 above)
[the remark this answers is not in the document]
Comment (sion-docs) [reply app/api/errors/IndividualCalculationsErrors.scala]: Revise to 'You are outside the amendment window'. This is for consistency with other error messages, and 'The request cannot be completed' is redundant as they are receiving an error message.

**Comment by hmrc-web-operations** [comment]
Thank you for submitting a pull request on this HMRC repository. A member of the HMRC team will review your changes and action accordingly.

HMRC reviewer: please reference [this runbook](https://github.com/hmrc/build-and-deploy/blob/main/runbooks/reviewing_external_pull_requests.md) for guidance on how to proceed.

**Review by GazalNebHMRC — APPROVED** [review]

**Pull request closed without merging**